tests/test_edgeql_insert.py

```diff
@@ -2264,43 +2264,113 @@ async def test_edgeql_insert_unless_conflict_01(self):
-                    INSERT test::Note {name := "hello"}
-                    UNLESS CONFLICT ON .name;
-                ''')
-
-        async with self._run_and_rollback():
-            with self.assertRaisesRegex(
-                    edgedb.QueryError,
-                    "UNLESS CONFLICT property must be a SINGLE property"):
-                await self.con.query(r'''
-                    INSERT test::Person {name := "hello", multi_prop := "lol"}
-                    UNLESS CONFLICT ON .multi_prop;
-                ''')
+                    UNLESS CONFLICT ON .name
+                    ELSE test::DefaultTest1
+                ) {name};
+            ''')
+
+        async with self.assertRaisesRegexTx(
+                edgedb.QueryError,
+                "possibly more than one element returned by an expression "
+                "for a computable link 'foo' declared as 'single'"):
+            await self.con.query(r'''
+                WITH MODULE test,
+                     X := (
+                        INSERT Person {name := "hello"}
+                        UNLESS CONFLICT ON .name
+                        ELSE (DETACHED Person)
+                    )
+                SELECT stdgraphql::Query {
+                    single foo := X
+                };
+            ''')
+
+        async with self.assertRaisesRegexTx(
+                edgedb.QueryError,
+                "possibly more than one element returned by an expression "
+                "for a computable link 'foo' declared as 'single'"):
+            await self.con.query(r'''
+                WITH MODULE test,
+                     X := (
+                        INSERT Person {name := "hello"}
+                        UNLESS CONFLICT ON .name
+                        ELSE Note
+                    )
+                SELECT stdgraphql::Query {
+                    single foo := X
+                };
+            ''')
+
+        async with self.assertRaisesRegexTx(
+                edgedb.QueryError,
+                "possibly an empty set returned by an expression for a "
+                "computable link 'foo' declared as 'required'"):
+            await self.con.query(r'''
+                WITH MODULE test,
+                     X := (
+                        INSERT Person {name := "hello"}
+                        UNLESS CONFLICT ON .name
+                    )
+                SELECT stdgraphql::Query {
+                    required foo := X
+                };
+            ''')
+
+        async with self.assertRaisesRegexTx(
+                edgedb.QueryError,
+                "possibly more than one element returned by an expression "
+                "for a computable link 'foo' declared as 'single'"):
+            await self.con.query(r'''
+                WITH MODULE test,
+                    X := (INSERT Person {
+                        name := "Test",
+                        case_name := "Foo",
+                    }
+                UNLESS CONFLICT ELSE Person),
```

Comment:
I see why `UNLESS CONFLICT ELSE SELECT` returns `MANY`, but I'm not 100% sure it's the best semantics.  Intuitively, any `INSERT` is expected to be `AT_MOST_ONE` and I see potential for confusion from errors raised in queries like this.

Perhaps we should rethink allowing `UNLESS CONFLICT ELSE` without an explicit conflict spec?  Postgres bans bare `ON CONFLICT DO UPDATE` for this reason.

Reply:
I dropped it for now